cpp/src/arrow/python/sequence.h

```diff
@@ -0,0 +1,137 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PYTHON_ARROW_SEQUENCE_H
+#define PYTHON_ARROW_SEQUENCE_H
```

Comment:
Any reason to think this would ever be used outside `python_to_arrow.cc`? I would suggest one of these options:

* Put both this and `DictBuilder` as inline definitions in `python_to_arrow-internal.h` and include in `python_to_arrow.cc`. Do not install this header. This even may improve performance due to better inlining
* Inline SequenceBuilder in python_to_arrow.cc before DictBuilder